File: frontend/lib/models/emergency.dart
class EmergencyModel {
  final String id;
  final String userId;
  final String type;
  final String status;
  final String? assignedOrgId;
  final String? assignedVolunteerId;
  final double locationLat;
  final double locationLng;
  final DateTime createdAt;
  final DateTime? updatedAt;

  EmergencyModel({
    required this.id,
    required this.userId,
    required this.type,
    required this.status,
    this.assignedOrgId,
    this.assignedVolunteerId,
    required this.locationLat,
    required this.locationLng,
    required this.createdAt,
    this.updatedAt,
  });

  factory EmergencyModel.fromJson(Map<String, dynamic> json) {
    return EmergencyModel(
      id: json['id'] ?? json['emergency_id'] ?? '',
      userId: json['user_id'] ?? '',
      type: json['type'] ?? '',
      status: json['status'] ?? 'pending',
      assignedOrgId: json['assigned_org_id'],
      assignedVolunteerId: json['assigned_volunteer_id'],
      locationLat: (json['location_lat'] ?? 0).toDouble(),
      locationLng: (json['location_lng'] ?? 0).toDouble(),
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'])
          : null,
    );
  }

  bool get isPending => status == 'pending';
  bool get isAccepted => status == 'accepted';
  bool get isCompleted => status == 'completed';

  String get typeIcon {
    switch (type) {
      case 'fire':
        return '🔥';
      case 'medical':
        return '🏥';
      case 'crime':
        return '🚨';
      default:
        return '⚠️';
    }
  }

  String get typeLabel {
    switch (type) {
      case 'fire':
        return 'Fire';
      case 'medical':
        return 'Medical';
      case 'crime':
        return 'Crime';
      default:
        return 'Emergency';
    }
  }
}

File: backend/app/models/emergency.py
import uuid
import enum
from datetime import datetime, timezone
from typing import Optional, TYPE_CHECKING

from sqlalchemy import Float, DateTime, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID as PG_UUID

from app.database import Base

if TYPE_CHECKING:
    from app.models.account import Account
    from app.models.organization import Organization


class EmergencyType(str, enum.Enum):
    FIRE = "fire"
    MEDICAL = "medical"
    ACCIDENT = "accident"
    NATURAL_DISASTER = "natural_disaster"
    CRIME = "crime"

    @classmethod
    def _missing_(cls, value):
        if isinstance(value, str):
            val_clean = value.strip().lower()
            for member in cls:
                if member.value == val_clean or member.name.lower() == val_clean:
                    return member
        return None


class EmergencyStatus(str, enum.Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    COMPLETED = "completed"
    CANCELLED = "cancelled"

    @classmethod
    def _missing_(cls, value):
        if isinstance(value, str):
            val_clean = value.strip().lower()
            for member in cls:
                if member.value == val_clean or member.name.lower() == val_clean:
                    return member
        return None


class RobustEmergencyTypeEnum(SAEnum):
    """
    Native PostgreSQL Enum for emergency_type_enum that binds correctly to PostgreSQL
    while overriding result_processor to safely handle case-insensitive values without LookupError.
    """
    def result_processor(self, dialect, coltype):
        def process(value):
            if value is None:
                return None
            if isinstance(value, self.enum_class):
                return value
            val_clean = str(value).strip().lower()
            for member in self.enum_class:
                if str(member.value).lower() == val_clean or member.name.lower() == val_clean:
                    return member
            try:
                return self.enum_class(val_clean)
            except Exception:
                return EmergencyType.MEDICAL
        return process

    def bind_processor(self, dialect):
        def process(value):
            if value is None:
                return None
            if isinstance(value, self.enum_class):
                return value.value
            val_clean = str(value).strip().lower()
            for member in self.enum_class:
                if str(member.value).lower() == val_clean or member.name.lower() == val_clean:
                    return member.value
            return val_clean
        return process


class RobustEmergencyStatusEnum(SAEnum):
    """
    Native PostgreSQL Enum for emergency_status_enum that binds correctly to PostgreSQL
    while overriding result_processor to safely handle case-insensitive values without LookupError.
    """
    def result_processor(self, dialect, coltype):
        def process(value):
            if value is None:
                return None
            if isinstance(value, self.enum_class):
                return value
            val_clean = str(value).strip().lower()
            for member in self.enum_class:
                if str(member.value).lower() == val_clean or member.name.lower() == val_clean:
                    return member
            try:
                return self.enum_class(val_clean)
            except Exception:
                return EmergencyStatus.PENDING
        return process

    def bind_processor(self, dialect):
        def process(value):
            if value is None:
                return None
            if isinstance(value, self.enum_class):
                return value.value
            val_clean = str(value).strip().lower()
            for member in self.enum_class:
                if str(member.value).lower() == val_clean or member.name.lower() == val_clean:
                    return member.value
            return val_clean
        return process


class Emergency(Base):
    """An SOS emergency event created by a user."""

    __tablename__ = "emergencies"

    id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("accounts.id"), nullable=False
    )
    type: Mapped[EmergencyType] = mapped_column(
        RobustEmergencyTypeEnum(EmergencyType, name="emergency_type_enum", values_callable=lambda x: [e.value for e in x]),
        nullable=False, index=True, default=EmergencyType.MEDICAL
    )
    status: Mapped[EmergencyStatus] = mapped_column(
        RobustEmergencyStatusEnum(EmergencyStatus, name="emergency_status_enum", values_callable=lambda x: [e.value for e in x]),
        default=EmergencyStatus.PENDING, index=True
    )
    assigned_org_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("organizations.account_id"), nullable=True, index=True
    )
    assigned_volunteer_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("volunteers.account_id"), nullable=True, index=True
    )
    location_lat: Mapped[float] = mapped_column(Float, nullable=False, index=True)
    location_lng: Mapped[float] = mapped_column(Float, nullable=False, index=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), index=True
    )
    updated_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=True,
    )

    user: Mapped["Account"] = relationship(foreign_keys=[user_id], lazy="selectin")
    assigned_org: Mapped[Optional["Organization"]] = relationship(foreign_keys=[assigned_org_id], lazy="selectin")
    assigned_volunteer: Mapped[Optional["Volunteer"]] = relationship(foreign_keys=[assigned_volunteer_id], lazy="selectin")
